requests-for-comments/2024-05-07-repeatable-surveys.md

```diff
@@ -0,0 +1,75 @@
+
+2. We see surveys created with a repeated schedule after this feature ships, I'm not sure of an exact number right now, but I assume that once we give customers the ability
+   to repeat a survey, they'll start using it.
+
+## Context
+_What are our competitors doing, what are the technical constraints, what are customers asking for, what does the data tell us, are there external motivations (e.g. launch week, enterprise contract)?_
+
+Our competitors have already built tools that help customers track user sentiment over a time period, allowing them to repeat surveys and gain insights into changing user behavior.
+
+Our current workarounds of `just repeat the survey manually` might get them to repeat the survey, but there's no way to gain any useful insights since the responses will end up in different surveys.
+
+### Similar tools
+1. [Pendo's custom NPS survey type](https://support.pendo.io/hc/en-us/articles/360033527151-Set-up-an-NPS-survey#h_01HGD0SKYSRB6SNKXBFJ1QG7M7)
+2. [Surveymonkey's multiple responses](https://help.surveymonkey.com/en/surveymonkey/send/allowing-multiple-responses/)
+
+
+## Design
+_What are the key user experience and technical design decisions / trade-offs?_
+
+A survey is assumed to have only one response in the system and we have filter conditions that govern **who** can see the survey.
+A user's survey response action is tracked as a boolean, i.e. we add a person property with the survey-id `$survey_response/<surveyID>` when a user submits or dismisses a review.
+
+To support this feature, we need to introduce the concept of `survey iterations` to the UI and person properties.
+
+Here's what we need to do to solve this : 
+
+### Data model changes to the surveys table
+1. Add properties to the survey data model to track 
+
+| Column name | type | description | nullable |
+|:----|:----|:----|:----|
+| `survey_frequency` | &lt;some cron indicator, could be varchar&gt; |  Frequency at which the survey should be repeated | Yes | 
+| `survey_max_iteration_count` | integer | max number of times the survey should be repeated |  Yes |
+
+This design allows our users to setup such surveys:
+
+1. `repeat this survey 4 times, once every 3 months`.
+2. They can also just leave the `survey_max_iteration_count` as zero to have an infinitely repeating survey `repeat this survey, once every 3 months`
+
+These two properties are nullable and we need to make our work backwards compatible to make existing survey responses function without issues to show NPS scores etc. 
+
+#### UI changes. 
+1. Show the user an interface to pick the repeat interval of the survey, this can be relative (every 4 months) or absolute (pick specific dates). 
+4. Show the user an interface to pick the iteration count, how many times should this survey repeat? 
+5. If a survey has an iteration count of more than 1, change the existing UI to run custom clickhouse queries to account for a survey with multiple response sets, one for each iteration. 
```

Comment:
would love to hear more about what this means, and how we want to do this?

For example, what if during the 2nd iteration, the same person never responds? Does it still make sense to breakdown results into iteration sets?

It's not clear to me how we should special case this, if at all.

I _think_ what might be useful here is a 2nd view only for users who've responded multiple times, and aggregating only for these users to show changes over time: so we have an overall view, that treats multiple responses as distinct, and this iteration-aggregation view, that only looks at people who've responded multiple times.


To make the above magic work, we absolutely need to make sure we get rid of this bug of users accidentally responding multiple times, which I think you're considering already.

Reply:
> For example, what if during the 2nd iteration, the same person never responds? 
Great callout, I hadn't considered variable response rate from users over a time period. I'll do some more thinking here and update the doc. 

> I think what might be useful here is a 2nd view only for users who've responded multiple times
Yeah, this makes sense. You can't track NPS if its a new set of people every time who responded in a given time period.

Reply:
I added some more color about how the results will be displayed and additional chart types we can show